Handle `read_chan_signer` in bindings (761 bindings updates)

TheBlueMatt opened the pull request:
Based on #761, this handles the new associated-type-returning trait method in the bindings. The bindings bits are decently large, so its going to stand as its own PR here. Between binidngs-update commits which build+run, a bindings diff commit exists to ease review, though those can be squashed before merge.

TheBlueMatt commented:
Rebased onto new upstream changes.

jkczyz commented on c-bindings-gen/src/types.rs:
```diff
@@ -211,6 +234,16 @@ impl<'a> GenericTypes<'a> {
 			}
+		} else {
+			let mut it = path.segments.iter();
+			if path.segments.len() == 2 && format!("{}", it.next().unwrap().ident) == "Self" {
```
Perhaps a one-line comment spelling out what case this handling would be useful.

jkczyz reviewed (commented):
Quite a few nice clean-ups in here!

TheBlueMatt commented on c-bindings-gen/src/types.rs:
```diff
@@ -193,6 +188,34 @@ impl<'a> GenericTypes<'a> {
+							assert_simple_bound(&tr);
+							if let Some(mut path) = types.maybe_resolve_path(&tr.path, None) {
+								if types.skip_path(&path) { continue; }
+								let new_ident = if path != "std::ops::Deref" {
```
I added a comment, but in general we handle `Deref<Target=X> as X` because they're mostly used for traits which we handle by `impl Deref<Target=Self>` for our traits.

TheBlueMatt commented:
Addressed review comments and rebased on latest upstream (without any changes except pulling the bindings update commit out to its own commit and redoing it for upstream changes).
